materials/section-2/README.md

```diff
@@ -0,0 +1,139 @@
+# Backend Programming
+
+In this section we will delve into backend programming with Ring and Compojure. Compojure is a lightweight routing library built on top of Ring which is an http library for Clojure backend development.
+
+Breakdown:
+
+- Ring -> Http server
+- Compojure -> Routing lib
+
+To scaffold a Compojure project, run `lein new compojure <name>` (replace name with a the name of your project). This command will create a Compojure project with the following folder structure:
+
+```
+<name>/
+├── README.md
+├── project.clj
+├── resources
+│   └── public
+├── src
+│   └── <name>
+│       └── handler.clj
+└── test
+    └── <name>
+        └── handler_test.clj
+```
+
+The most interesting files are without a doubt, `project.clj` and `src/<name>/handler.clj`. The `project.clj` file is the file that describes the project to Leiningen. The project dependencies, name, description etc. are describes inside the `project.clj` file. Incase of a Compojure project, it also defines the entry point for the web-server: `:ring {:handler <name>.handler/app}`. This tells Ring where to look for the entrypoint so it can bootstrap the server. The second interesting file, is the `handler.clj` file. This is where the application is bootstrapped. It will look something like this:
+
+```clojure
+(ns <name>.handler
+  (:require [compojure.core :refer :all]
+            [compojure.route :as route]
+            [ring.middleware.defaults :refer [wrap-defaults site-defaults]]))
+
+(defroutes app-routes
+  (GET "/" [] "Hello World")
+  (route/not-found "Not Found"))
+
+(def app
+  (wrap-defaults app-routes site-defaults))
+```
+
+Small breakdown:
+
+```clojure
+(ns <name>.handler
+  (:require [compojure.core :refer :all]
+            [compojure.route :as route]
+            [ring.middleware.defaults :refer [wrap-defaults site-defaults]]))
+```
+
+Namespace is defined and the core of Compojure is included along with the route namespace. It also requires the middleware namespace from Ring. Middlewares are just an easy way to handle requests before they reach your route handler. Ring has a few default ones which Compojure uses when it bootstraps the project.
+
+```clojure
+(defroutes app-routes
+  (GET "/" [] "Hello World")
+  (route/not-found "Not Found"))
+```
+
+`defroutes` is a macro provided by Compojure. It takes two parameters. A name and route handlers. A route handler is nothing more than a basic function. Compojure provides a handful of route handlers. `GET`, `POST`, `PATCH`, `PUT`, `HEAD`, `DELETE`, `OPTIONS` and `ANY`.
+All of these route handlers work in the same way. Each of them take a path as the first argument, a vector with bound route parameters, and a route body as the last argument.
```

Comment:
Could be written a bit clearer. A vector is not necessarily needed as the second argument. The vector is actually a shorthand for destructuring the basic request map that gets passed in